barretenberg/cpp/src/barretenberg/vm/avm_trace/avm_trace.cpp

```diff
@@ -155,19 +198,19 @@ void AvmTraceBuilder::op_add(
         .main_clk = clk,
         .main_alu_in_tag = FF(static_cast<uint32_t>(in_tag)),
         .main_call_ptr = call_ptr,
-        .main_ia = a,
-        .main_ib = b,
-        .main_ic = c,
-        .main_ind_a = res.indirect_flag_a ? FF(a_offset) : FF(0),
-        .main_ind_b = res.indirect_flag_b ? FF(b_offset) : FF(0),
-        .main_ind_c = res.indirect_flag_c ? FF(dst_offset) : FF(0),
-        .main_ind_op_a = FF(static_cast<uint32_t>(res.indirect_flag_a)),
-        .main_ind_op_b = FF(static_cast<uint32_t>(res.indirect_flag_b)),
-        .main_ind_op_c = FF(static_cast<uint32_t>(res.indirect_flag_c)),
+        .main_ia = read_a.val,
+        .main_ib = read_b.val,
+        .main_ic = write_c.val,
+        .main_ind_a = FF(read_a.indirect_address),
+        .main_ind_b = FF(read_b.indirect_address),
+        .main_ind_c = FF(write_c.indirect_address),
+        .main_ind_op_a = FF(static_cast<uint32_t>(read_a.is_indirect)),
+        .main_ind_op_b = FF(static_cast<uint32_t>(read_b.is_indirect)),
+        .main_ind_op_c = FF(static_cast<uint32_t>(write_c.is_indirect)),
         .main_internal_return_ptr = FF(internal_return_ptr),
-        .main_mem_idx_a = FF(res.direct_a_offset),
-        .main_mem_idx_b = FF(res.direct_b_offset),
-        .main_mem_idx_c = FF(res.direct_c_offset),
+        .main_mem_idx_a = FF(read_a.direct_address),
+        .main_mem_idx_b = FF(read_b.direct_address),
+        .main_mem_idx_c = FF(write_c.direct_address),
```

Comment:
In another PR, we should introduce a function which sets all the memory-related fields in the Row based on a vector of MemOp.